### PR TITLE
[FEAT] Add lambda closure to manual_optimizer_step

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds the possibility for the user to add its own lambda_closure.

Resolving #4485

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by williamFalcon — APPROVED** [review]

**Comment by SeanNaren** [comment]
@tchaton Could you expose the args/kwargs also within the accelerator optimizer steps? There are a few places where the optimizer_step is overridden in our acceleartors.

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -127,7 +127,9 @@ def optimizer_step(self, optimizer, batch_idx, opt_idx, lambda_closure):
             optimizer_closure=lambda_closure,
             on_tpu=False,  # TPUAccelerator class sets this as True
             using_native_amp=using_native_amp,
-            using_lbfgs=is_lbfgs
+            using_lbfgs=is_lbfgs,
+            *args,
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:131]: personally, I would not use it... but should be fine :]

File: .gitignore [hunk 2]
```diff
@@ -140,3 +140,4 @@ mlruns/
 pytorch\ lightning
 test-reports/
 wandb
+.forked/
```
Comment (Borda) [inline .gitignore:143]: how does this folder happen?

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -1106,7 +1107,12 @@ def training_step(...):
         self.trainer.train_loop.backward(loss, optimizer, -1, *args, **kwargs)
         self._running_manual_backward = False
 
-    def manual_optimizer_step(self, optimizer: Optimizer, force_optimizer_step:bool = False) -> None:
+    def manual_optimizer_step(self,
+                              optimizer: Optimizer,
+                              *args,
+                              make_optimizer_step: Optional[bool] = None,
+                              optimizer_closure: Optional[Callable] = None,
+                              ** kwargs) -> None:
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1115]:
```suggestion
def manual_optimizer_step(
            self,
            optimizer: Optimizer,
            *args,
            make_optimizer_step: Optional[bool] = None,
            optimizer_closure: Optional[Callable] = None,
            **kwargs
    ) -> None:
```
just formatting

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1106,7 +1107,12 @@ def training_step(...):
         self.trainer.train_loop.backward(loss, optimizer, -1, *args, **kwargs)
         self._running_manual_backward = False
 
-    def manual_optimizer_step(self, optimizer: Optimizer, force_optimizer_step:bool = False) -> None:
+    def manual_optimizer_step(self,
+                              optimizer: Optimizer,
+                              *args,
+                              make_optimizer_step: Optional[bool] = None,
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1113]: What None for optional bool means, also not mentioned docs
```suggestion
                              make_optimizer_step: bool = False,
```

File: pytorch_lightning/core/lightning.py [hunk 5]
```diff
@@ -1116,36 +1122,132 @@ def manual_optimizer_step(self, optimizer: Optimizer, force_optimizer_step:bool
...
+                        opt_dis,
+                        optimizer_closure=dis_closure,
+                        make_optimizer_step=True)
         """
         # make sure we're using manual opt
         self._verify_is_manual_optimization('manual_optimizer_step')
 
-        if not self.trainer.train_loop.should_accumulate() or force_optimizer_step:
+        should_make_optimizer_step = not self.trainer.train_loop.should_accumulate()
+        make_optimizer_step = make_optimizer_step if make_optimizer_step is not None else should_make_optimizer_step
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1222]:
```suggestion
        # take option from `manual_optimizer_step` if any explicit given, otherwise follow Trainer setting
        make_optimizer_step = make_optimizer_step if make_optimizer_step is not None else should_make_optimizer_step
```

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -1116,36 +1122,132 @@ def manual_optimizer_step(self, optimizer: Optimizer, force_optimizer_step:bool
...
 
-            self.trainer.train_loop.optimizer_step(optimizer, None, self.trainer.batch_idx, mock_optimizer_closure)
+            optimizer_closure = optimizer_closure if isinstance(optimizer_closure, types.FunctionType) else mock_optimizer_closure
+
+            self.trainer.train_loop.optimizer_step(optimizer,
+                                                   None,
+                                                   self.trainer.batch_idx,
+                                                   optimizer_closure,
+                                                   *args,
+                                                   **kwargs)
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1237]:
```suggestion
            self.trainer.train_loop.optimizer_step(
                optimizer,
                None,
                self.trainer.batch_idx,
                optimizer_closure,
                *args,
                **kwargs,
            )
```
just formatting

File: pytorch_lightning/core/lightning.py [hunk 7]
```diff
@@ -1116,36 +1122,132 @@ def manual_optimizer_step(self, optimizer: Optimizer, force_optimizer_step:bool
...
+        make_optimizer_step = make_optimizer_step if make_optimizer_step is not None else should_make_optimizer_step
+
+        if make_optimizer_step:
 
             # mock closure function as the user is responsible to call `manual_backward`
             def mock_optimizer_closure():
                 return
 
-            self.trainer.train_loop.optimizer_step(optimizer, None, self.trainer.batch_idx, mock_optimizer_closure)
+            optimizer_closure = optimizer_closure if isinstance(optimizer_closure, types.FunctionType) else mock_optimizer_closure
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1230]: maybe rename the moct_ to empty_? just an idea
btw, what about using type Callable?

File: tests/trainer/optimization/test_manual_optimization.py [hunk 8]
```diff
@@ -621,3 +623,277 @@ def configure_optimizers(self):
...
     num_manual_backward_calls = 3
     assert trainer.dev_debugger.count_events('backward_call') == limit_train_batches * num_manual_backward_calls
+
+
+def test_manual_optimizer_step_with_optimizer_closure(tmpdir):
+    os.environ['PL_DEV_DEBUG'] = '1'
+
+    """
+    Tests that `manual_optimizer_step` works with optimizer_closure
+    """
```
Comment (Borda) [inline tests/trainer/optimization/test_manual_optimization.py:633]:
```suggestion
    """
    Tests that `manual_optimizer_step` works with optimizer_closure
    """
    os.environ['PL_DEV_DEBUG'] = '1'

```

File: tests/trainer/optimization/test_manual_optimization.py [hunk 9]
```diff
@@ -621,3 +623,277 @@ def configure_optimizers(self):
...
+    assert trainer.logger_connector.progress_bar_metrics["train_loss_step"] == model._losses[-1]
+    assert trainer.logger_connector.progress_bar_metrics["train_loss_epoch"] == torch.stack(model._losses).mean()
+
+
+def test_manual_optimizer_step_with_optimizer_closure_and_accumulated_grad(tmpdir):
+    os.environ['PL_DEV_DEBUG'] = '1'
+
+    """
+    Tests that `manual_optimizer_step` works with optimizer_closure and accumulated_grad
+    """
```
Comment (Borda) [inline tests/trainer/optimization/test_manual_optimization.py:710]:
```suggestion
    """
    Tests that `manual_optimizer_step` works with optimizer_closure and accumulated_grad
    """
    os.environ['PL_DEV_DEBUG'] = '1'

```

File: tests/trainer/optimization/test_manual_optimization.py [hunk 10]
```diff
@@ -621,3 +623,277 @@ def configure_optimizers(self):
...
+    assert trainer.dev_debugger.count_events('backward_call') == limit_train_batches * 2
+
+
+@patch("torch.optim.SGD.step")
+def test_manual_optimizer_step_with_optimizer_closure_and_extra_arguments(step_mock, tmpdir):
+    os.environ['PL_DEV_DEBUG'] = '1'
+
+    """
+    Tests that `manual_optimizer_step` works with optimizer_closure and extra arguments
+    """
```
Comment (Borda) [inline tests/trainer/optimization/test_manual_optimization.py:770]:
```suggestion
    """
    Tests that `manual_optimizer_step` works with optimizer_closure and extra arguments
    """
    os.environ['PL_DEV_DEBUG'] = '1'

```

File: tests/trainer/optimization/test_manual_optimization.py [hunk 11]
```diff
@@ -621,3 +623,277 @@ def configure_optimizers(self):
...
+
+
+@patch("torch.optim.Adam.step")
+@patch("torch.optim.SGD.step")
+def test_manual_optimizer_step_with_optimizer_closure_with_different_frequencies(mock_sgd_step, mock_adam_step, tmpdir):
+    os.environ['PL_DEV_DEBUG'] = '1'
+
+    """
+    Tests that `manual_optimizer_step` works with optimizer_closure and different accumulated_gradient frequency
+    """
```
Comment (Borda) [inline tests/trainer/optimization/test_manual_optimization.py:824]:
```suggestion
    """
    Tests that `manual_optimizer_step` works with optimizer_closure and different accumulated_gradient frequency
    """
    os.environ['PL_DEV_DEBUG'] = '1'

```

**Review by Borda — APPROVED** [review]
LGTM, just still reading the tests... :]

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/accelerators/accelerator.py:131]: I followed Will advice. But it makes sense to carry arguments over.

File: pytorch_lightning/core/lightning.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1113]: If nothing is provided, it will use accumulated_grad_batches. I added this several times in the examples.

File: .gitignore (hunk 2 above)
Comment (tchaton) [reply .gitignore:143]: Conversion for virtualenvs: .venv for main repo, .forked for personal forked one.

File: pytorch_lightning/core/lightning.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1230]: I have seen several times the right way was `types.FunctionType`
